### DISCUSSION TITLE
Stream Classes

### DISCUSSION THREAD

**Discussion by ryanmrichard** [NWChemEx/ParallelZone#6] [Ideas]
The stream classes will be used for logging, output, and error messages. I'd recommend an API that's basically the same as `std::ostream` with maybe a `printf` function for those who prefer to use that sort of API. 

Requirements:
-------------------

- Interoperability with `std::ostream` / `std::istream`
  - This is the C++ standard library native way of printing
  - Many existing C++ libraries assume this
  - Makes sure we can treat stdout and files with a common API
- Parallel aware
  - Will want to have only certain ranks print
  - Would be nice if the equivalent of `std::cout<< my_rank << " " << my_data << std::endl;` didn't produce mangled garbage
- Intelligent floating point printing (dump them in full-precision automatically)
- Asynchronous (particularly file writing)
  - The operating system may help with this already

Great to have:
-------------------
- `printf` API; it's sometimes much easier than streams
- color

**Comment by keceli** [comment]
I guess, we might prefer using a library if we can find a good one. So, I checked out some and listed ones with compatible licenses and active development. All needs to be configured for rank 0 print, but they guarantee thread safety.

1. [boost::log](https://www.boost.org/doc/libs/1_76_0/libs/log/doc/html/index.html): Might have more overhead compared to other libraries, but has more reliable support.
2. [Easylogging++](https://github.com/amrayn/easyloggingpp): Has macros for performance tracking and stack tracing.
3. [Quill](https://github.com/odygrd/quill): Uses {fmt} library to support Python style API and STL types/containers.
4. [Binlog](https://github.com/morganstanley/binlog): Requires conversion from binary to text, so not suitable for our purposes, but could be the right logger for high throughput calculations. 
5. [Plog](https://github.com/SergiusTheBest/plog) Might be slower than others since it is synchronous, but looks much simpler to use and extend.

**Reply by ryanmrichard** [reply]
What are the features these libraries bring to the table? Writing classes that wrap streams isn't too hard and if none of them support parallel printing, it sounds like we're going to have to do it anyways.

**Comment by twindus** [comment]
Thanks for looking into this @keceli.  Can you comment on the compatibility with std::ostream?  I guess one possible advantage of boost::log is that we already have boost dependencies?
I can't really speak to the other libraries, but it would be nice if the printing could be asynchronous, so maybe Plog won't work?

**Reply by keceli** [reply]
Easylogging++ has a base Loggable class that implements pure-virtual function `void log(std::ostream&) const`, so I think it is compatible. With boost::log, it is possible to redirect `ostream` to logs by defining a "sink" as described [here](https://stackoverflow.com/questions/30958520/how-to-redirect-ostream-to-boost-log-library).
